message/infrastructure/sql/schema_adapter.go

```diff
@@ -0,0 +1,122 @@
+package sql
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/pkg/errors"
+)
+
+// SchemaAdapter produces the SQL queries and arguments appropriately for a specific schema and dialect
+// It also transforms sql.Rows into Watermill messages.
+type SchemaAdapter interface {
+	// InsertQuery returns the SQL query that will insert the Watermill message into the SQL storage.
+	InsertQuery(topic string, msgs message.Messages) (string, []interface{}, error)
+
+	// SelectQuery returns the SQL query that returns the next unread message for a given consumer group.
+	SelectQuery(topic string, consumerGroup string, offsetsAdapter OffsetsAdapter) (string, []interface{})
+
+	// UnmarshalMessage transforms the Row obtained from the SQL query into a Watermill message.
+	// It also returns the offset of the last read message, for the purpose of acking.
+	UnmarshalMessage(row *sql.Row) (offset int, msg *message.Message, err error)
+
+	// SchemaInitializingQueries returns SQL queries which will make sure (CREATE IF NOT EXISTS)
+	// that the appropriate tables exist to write messages to the given topic.
+	SchemaInitializingQueries(topic string) []string
+}
+
+// DefaultSchema is a default implementation of SchemaAdapter that works with the following schema:
+//
+// `offset` BIGINT NOT NULL AUTO_INCREMENT,
+// `uuid` VARCHAR(36) NOT NULL,
+// `created_at` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+// `payload` JSON DEFAULT NULL,
+// `metadata` JSON DEFAULT NULL,
+type DefaultSchema struct {
+	// GenerateMessagesTableName may be used to override how the messages table name is generated.
+	GenerateMessagesTableName func(topic string) string
+}
+
+func (s DefaultSchema) SchemaInitializingQueries(topic string) []string {
+	createMessagesTable := strings.Join([]string{
+		"CREATE TABLE IF NOT EXISTS " + s.MessagesTable(topic) + " (",
+		"`offset` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,",
+		"`uuid` VARCHAR(36) NOT NULL,",
+		"`created_at` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,",
+		"`payload` JSON DEFAULT NULL,",
+		"`metadata` JSON DEFAULT NULL",
+		");",
+	}, "\n")
+
+	return []string{createMessagesTable}
+}
+
+func (s DefaultSchema) InsertQuery(topic string, msgs message.Messages) (string, []interface{}, error) {
```

Comment:
I added support for multi insert with one query, it should be much faster

Reply:
Each message adds three args to the slice. I don't think I'm worried about the golang slice limits, but maybe we could write a test to see how many messages the sql driver can handle and if something stupid and unexpected happens with some relatively sane number.

Reply:
If there's a panic for example, add some safeguards.

Reply:
it's not our domain (like addresses, validation you remember :P)